modules/channel-web/src/backend/db.ts

```diff
@@ -10,14 +9,79 @@ import { Config } from '../config'
 export default class WebchatDb {
   knex: any
   users: typeof sdk.users
+  private queued_messages: any[] = []
+  private queued_convos: Dic<any> = {}
+  private message_lock: boolean
+  private convo_lock: boolean
 
   constructor(private bp: typeof sdk) {
     this.users = bp.users
     this.knex = bp['database'] // TODO Fixme
+    setInterval(() => this.flushMessages(), ms('5s'))
+    setInterval(() => this.flushConvoUpdates(), ms('5s'))
   }
 
-  async getUserInfo(userId) {
-    const { result: user } = await this.users.getOrCreateUser('web', userId)
+  async flushMessages() {
+    if (this.message_lock || !this.queued_messages.length) {
+      return
+    }
+    this.message_lock = true
+
+    try {
+      const original = this.queued_messages
+      this.queued_messages = []
+
+      this.knex('web_messages')
+        .insert(original)
+        .catch(() => {
+          this.queued_messages = original.concat(this.queued_messages)
+        })
+    } finally {
+      this.message_lock = false
```

Comment:
Instead of using a message lock, you can use the promise that you get from knex, there's an example here: https://github.com/botpress/botpress/blob/master/src/bp/core/services/middleware/event-collector.ts#L116

Prefer the spread operator instead of concat

Also, I think messages should be batched in smaller size, because there's a limit on the number of variables that can be filled in sql queries (999)

Reply:
When testing knex seems to be doing alright inserting 2500+ at the same time. Maybe I can make batch intervals faster.

I don't think it would be a good idea to accumulate messages in the queue if we were to set a maximum amount of inserts per sec.

Reply:
SQLite has a hard limit on 999 variables, so it must be split in small batches to make sure there's no error. Usually there's fluctuation in usage so the server can keep up with messages